.github/workflows/ci.yml

```diff
@@ -7,51 +7,63 @@ on:
   pull_request:
     branches: [ main ]
 
+env:
+  EXERCISES_DIRECTORY: ./exercises
+
 jobs:
-  validate_course:
+  list-exercises:
     runs-on: ubuntu-latest
-
+    outputs:
+      exercises: ${{steps.list.outputs.exercises}}
     steps:
+      - uses: actions/checkout@v3
+      - id: list
+        run: echo "::set-output name=exercises::$(cd $EXERCISES_DIRECTORY; ls |grep exercise_ | jq -cnR '[inputs | select(length>0)]')"
 
-      - name: Checkout Course Repo
-        uses: actions/checkout@v2
-        with:
-          path: course-repo
-          # fetch-depth: 0 indicates all history. (CMT requires them)
-          fetch-depth: 0
-        
-      - name: Setup Course Management Tools
-        uses: robinraju/release-downloader@v1.1
-        with:
-          repository: eloots/course-management-tools
-          tag: "1.0.0"
-          fileName: "course-management-tools.zip"
-          out-file-path: "."
-      - run: |
-          unzip course-management-tools.zip
-          echo "$GITHUB_WORKSPACE/course-management-tools/bin" >> $GITHUB_PATH
-      
+  validate_course:
+    runs-on: ubuntu-latest
+    needs: list-exercises
+    strategy:
+      fail-fast: false
+      matrix:
+        exercise: ${{fromJson(needs.list-exercises.outputs.exercises)}}   
+    steps:
+      - uses: actions/checkout@v3
       - name: Set up JDK 11
-        uses: actions/setup-java@v1
+        uses: actions/setup-java@v3
```

Comment:
Just a heads up `setup-java` has built-in sbt caching. You can just add a `cache: 'sbt'` under your `java-version`.

Reply:
Thanks for the tip. Implemented.